src/coreclr/jit/importercalls.cpp

```diff
@@ -4112,22 +4112,31 @@ GenTree* Compiler::impIntrinsic(CORINFO_CLASS_HANDLE    clsHnd,
             }
 #endif // defined(TARGET_ARM64) || defined(TARGET_RISCV64)
 
-#if defined(TARGET_XARCH) || defined(TARGET_ARM64) || defined(TARGET_RISCV64)
+#if defined(TARGET_64BIT) || defined(TARGET_X86)
             // TODO-ARM-CQ: reenable treating InterlockedCmpXchg32 operation as intrinsic
             case NI_System_Threading_Interlocked_CompareExchange:
             {
+#if defined(TARGET_LOONGARCH64)
+                // GT_CMPXCHG is NYI in lsraloongarch64
+                mustExpand = false;
+                break;
+#endif
+
                 var_types retType = JITtype2varType(sig->retType);
 
                 if (genTypeSize(retType) > TARGET_POINTER_SIZE)
                 {
                     break;
                 }
-#if !defined(TARGET_XARCH) && !defined(TARGET_ARM64)
+#if defined(TARGET_RISCV64)
+                // CAS for small types is natively supported starting with the Zabha extension; however,
+                // hardware support for RVA23 profile is not available at the time of writing.
                 else if (genTypeSize(retType) < 4)
                 {
+                    mustExpand = false;
```

Comment:
You cannot just choose to not expand a must-expand intrinsic. That's the whole point of must-expand.

It seems like the C# implementation should be adjusted to avoid this being a must-expand intrinsic when not supported.

Reply:
This is specific to AOT ABI, crossgen2 and corehost form factors seem to be fine without this patch. It seems we do alter mustExpand in the AOT block above https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/coreclr/jit/importercalls.cpp#L3390

Reply:
Yes, on NAOT certain more intrinsics are must-expand. That's not a problem.

Not expanding a must-expand intrinsic is a problem. Must expand means that the C# implementation is a self-recursive call. It is going to result in hang at runtime to not expand this.

Reply:
It could be that caller of https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs#L495 will run into recursion from line 501 because ILC rewrites generic to non-generic: https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/coreclr/tools/Common/TypeSystem/IL/Stubs/InterlockedIntrinsics.cs#L57

Reply:
BTW, I wasn't able to reproduce it locally. Perhaps @hez2010 could confirm this but it seems like godbolt is using x64 runtime libs (specifically corelib) for all ilc --targetarch values? e..g --targetarch arm (32-bit) fails differently than riscv64 https://godbolt.org/z/996aMTdE1.

Reply:
For ilc godbolt is using x64 corelibs for all archs because setting up the build of aotsdk for other archs is not trivial.

Reply:
Thanks for the clarification. The code @jakobbotsch pointed out lead to the theory that it's arch mismatch, glad that the mystery is resolved. :)

> We could potentially improve the godbolt dotnet build script to support aotsdk cross-build for naot: [compiler-explorer/dotnet-builder@41570b2/build/build.sh#L97-L102](https://github.com/compiler-explorer/dotnet-builder/blob/41570b24c82d4437c190d959097408c71e090e38/build/build.sh#L97-L102)

It sounds like we could build `clr.nativeaotlibs` subset for all desktop platforms (except linux-x86), but I'm not familiar with their build budget as it would likely increase the build time in a significant way.

Reply:
> but I'm not familiar with their build budget as it would likely increase the build time in a significant way

I don't think you need to concern about the build budget. There are already many compilers that require much more build time than dotnet. See https://github.com/compiler-explorer/compiler-workflows/actions, where clang/llvm takes more than an hour to build everyday.